blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/EnumElasticIO.java

```diff
@@ -0,0 +1,215 @@
+import org.elasticsearch.client.Client;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.search.SearchHit;
+
+import java.io.IOException;
+import java.util.*;
+
+import static org.elasticsearch.index.query.QueryBuilders.*;
+
+public class EnumElasticIO implements DiscoveryIO {
+
+    public static String ENUMS_INDEX_NAME_WRITE = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_ENUMS_INDEX_NAME_WRITE);
+    public static String ENUMS_INDEX_NAME_READ = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_INDEX_NAME_READ);
+    public static final String ENUMS_DOCUMENT_TYPE = "metrics";
+
+    private Client client;
+    private final Timer searchTimer = Metrics.timer(EnumElasticIO.class, "Search Duration");
+    private final Timer writeTimer = Metrics.timer(EnumElasticIO.class, "Write Duration");
+    private final Histogram batchHistogram = Metrics.histogram(EnumElasticIO.class, "Batch Sizes");
+    private Meter classCastExceptionMeter = Metrics.meter(EnumElasticIO.class, "Failed Cast to IMetric");
+    private Histogram queryBatchHistogram = Metrics.histogram(EnumElasticIO.class, "Query Batch Size");
+
+    public EnumElasticIO() {
+        this(RemoteElasticSearchServer.getInstance());
+    }
+
+    public EnumElasticIO(Client client) {
+        this.client = client;
+    }
+
+    public EnumElasticIO(ElasticClientManager manager) {
+        this(manager.getClient());
+    }
+
+    public void insertDiscovery(List<IMetric> batch) throws IOException {
+        batchHistogram.update(batch.size());
+        if (batch.size() == 0) {
+            return;
+        }
+
+        // TODO: check bulk insert result and retry
+        Timer.Context ctx = writeTimer.time();
+        try {
+            BulkRequestBuilder bulk = client.prepareBulk();
+            for (Object obj : batch) {
+                if (!(obj instanceof PreaggregatedMetric)) {
+                    classCastExceptionMeter.mark();
+                    continue;
+                }
+
+                // get PreaggregatedMetric
+                PreaggregatedMetric metric = (PreaggregatedMetric)obj;
+
+                // get rollup object of metric values
+                BluefloodEnumRollup rollup = (BluefloodEnumRollup) metric.getMetricValue();
+                if (rollup == null) {
+                    continue;
+                }
+
+                Locator locator = metric.getLocator();
+                Discovery md = new Discovery(locator.getTenantId(), locator.getMetricName());
+
+                Map<String, Object> info = new HashMap<String, Object>();
+
+                ArrayList<String> enumValues = rollup.getStringEnumValues();
+                if (enumValues.size() > 0) {
+                    info.put(ESFieldLabel.enum_values.toString(), enumValues);
+                }
+
+                md.withAnnotation(info);
```

Comment:
annotation are actually "extra fields" that are not either the metric_name or tenant_id in the Discovery class.  The way the Discovery class is setup is that any fields you would like to include in the index to be written to ES (such as "unit"), you would add it as annotation.

See the code here in the extracted Discovery class (formerly and inner class of ElasticIO), in method createSourceContent(), to see how the annotations is retrieved to prepare the json source for the index:  https://github.com/VinnyQ/blueflood/blob/3d6d289712cfcfb84187a5a069632aad08ba8300/blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/Discovery.java#L70-L72